core/src/test/java/org/infinispan/stream/DistributedWriteBehindStreamIteratorTest.java

```diff
@@ -0,0 +1,134 @@
+      super(false, null);
+   }
+
+   DistributedWriteBehindStreamIteratorTest async(boolean asyncStore) {
+      this.asyncStore = asyncStore;
+      return this;
+   }
+
+   DistributedWriteBehindStreamIteratorTest shared(boolean sharedStore) {
+      this.sharedStore = sharedStore;
+      return this;
+   }
+
+   @Override
+   public Object[] factory() {
+      return new Object[]{
+            new DistributedWriteBehindStreamIteratorTest().async(true).shared(true).cacheMode(CacheMode.REPL_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(false).shared(true).cacheMode(CacheMode.REPL_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(true).shared(false).cacheMode(CacheMode.REPL_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(false).shared(false).cacheMode(CacheMode.REPL_SYNC),
+
+            new DistributedWriteBehindStreamIteratorTest().async(true).shared(true).cacheMode(CacheMode.DIST_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(false).shared(true).cacheMode(CacheMode.DIST_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(true).shared(false).cacheMode(CacheMode.DIST_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(false).shared(false).cacheMode(CacheMode.DIST_SYNC),
+      };
+   }
+
+   @Override
+   protected Object[] parameterValues() {
+      return concat(super.parameterValues(), asyncStore, sharedStore);
+   }
+
+   @Override
+   protected String[] parameterNames() {
+      return concat(super.parameterNames(), "asyncStore", "sharedStore");
+   }
+
+   @Override
+   protected void enhanceConfiguration(ConfigurationBuilder builder) {
+      // We don't want to mess with the segments in the parent class
+      builder.clustering().hash().numSegments(60).consistentHashFactory(null);
+      DummyInMemoryStoreConfigurationBuilder dimscb = builder.persistence().addStore(DummyInMemoryStoreConfigurationBuilder.class);
+      if (sharedStore) {
+         dimscb.shared(true);
+      }
+      if (asyncStore) {
+         dimscb.storeName(getTestName())
+               .async().enable();
+      }
+   }
+
+   @Test
+   public void testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStoreWithReHash() throws Exception {
+      testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStore(true);
+   }
+
+   @Test
+   public void testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStoreWithoutRehash() throws Exception {
+      testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStore(false);
+   }
+
+   protected void testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStore(boolean rehashAware) {
+      Cache<Object, String> cache0 = cache(0, CACHE_NAME);
+
+      ClusterStreamManager clusterStreamManager = replaceWithSpy(cache0);
+
+      int amountToInsert = 200;
+      IntStream.range(0, amountToInsert).boxed().forEach(i -> cache0.put(i, i.toString()));
```

Comment:
Why so many keys and segments @wburns?

Reply:
Well the 3 segments in the original didn't work, because the originator had to be a backup owner of a given node for it to work. I can reduce them a bit more if you wanted, abut this didn't seem like too many to make sure the test always passed.

Reply:
Surely all the segments in the parent have a backup owner as well?

Reply:
I can try messing with it, just seemed easier to just increase the segments at the time :)